rust/feature-flags/src/flag_matching.rs

```diff
@@ -32,73 +166,288 @@ pub struct Person {
+    ) -> Result<(), FlagError> {
+        if self.database_client.is_none() {
+            error!("Database client is None");
+            return Err(FlagError::DatabaseUnavailable);
+        }
+
+        let mut conn = self
+            .database_client
+            .as_ref()
+            .expect("client should exist here")
+            .get_connection()
+            .await?;
+
+        let query = r#"
+            SELECT 
+                (SELECT "posthog_person"."properties"
+                 FROM "posthog_person"
+                 INNER JOIN "posthog_persondistinctid" 
+                 ON ("posthog_person"."id" = "posthog_persondistinctid"."person_id")
+                 WHERE ("posthog_persondistinctid"."distinct_id" = $1
+                        AND "posthog_persondistinctid"."team_id" = $2
+                        AND "posthog_person"."team_id" = $2)
+                 LIMIT 1) as person_properties,
+                
+                (SELECT json_object_agg("posthog_group"."group_type_index", "posthog_group"."group_properties")
+                 FROM "posthog_group"
+                 WHERE ("posthog_group"."team_id" = $2
+                        AND "posthog_group"."group_type_index" = ANY($3))) as group_properties
+        "#;
+
+        let group_type_indexes_vec: Vec<GroupTypeIndex> =
+            group_type_indexes.iter().cloned().collect();
+
+        let row: (Option<Value>, Option<Value>) = sqlx::query_as(query)
+            .bind(&self.distinct_id)
+            .bind(self.team_id)
+            .bind(&group_type_indexes_vec)
+            .fetch_optional(&mut *conn)
+            .await?
+            .unwrap_or((None, None));
+
+        let mut cache = self.properties_cache.write().await;
+
+        if let Some(person_props) = row.0 {
+            cache.person_properties = Some(
+                person_props
+                    .as_object()
+                    .unwrap()
+                    .iter()
+                    .map(|(k, v)| (k.clone(), v.clone()))
+                    .collect(),
+            );
+        }
+
+        if let Some(group_props) = row.1 {
+            let group_props_map: HashMap<GroupTypeIndex, HashMap<String, Value>> = group_props
+                .as_object()
+                .unwrap()
+                .iter()
+                .map(|(k, v)| {
+                    let group_type_index = k.parse().unwrap();
+                    let properties: HashMap<String, Value> = v
+                        .as_object()
+                        .unwrap()
+                        .iter()
+                        .map(|(k, v)| (k.clone(), v.clone()))
+                        .collect();
+                    (group_type_index, properties)
+                })
+                .collect();
+
+            cache.group_properties.extend(group_props_map);
+        }
+
+        Ok(())
+    }
+
+    async fn fetch_person_properties_from_db(&self) -> Result<HashMap<String, Value>, FlagError> {
```

Comment:
I kept this method around because I still want to be able to fall back in individual condition matching requests in the worst case scenario where the cache somehow doesn't exist; I'd rather be slow than fail.

Reply:
the cache here is in memory right? Not sure what you mean 🤔

Reply:
sorry, I feel like "cache somehow doesn't exist" was a confusing way of explaining this.  Basically, I have this method
```
    async fn get_person_properties(
        &self,
        flag_property_filters: &[PropertyFilter],
    ) -> Result<HashMap<String, Value>, FlagError> {
        if let Some(overrides) = can_compute_property_overrides_locally(
            &self.person_property_overrides,
            flag_property_filters,
        ) {
            return Ok(overrides);
        }

        let cache = self.properties_cache.read().await;
        if let Some(properties) = cache.person_properties.clone() {
            return Ok(properties);
        }
        drop(cache);

        let db_properties = self.fetch_person_properties_from_db().await?;
        let mut cache = self.properties_cache.write().await;
        cache.person_properties = Some(db_properties.clone());

        Ok(db_properties)
    }
    ```
that fetch person properties from (1) overrides, (2) the in-memory cache, and then (3) from Postgres.  With my current implementation, I was loading up the properties cache at the start of the request so that any subsequent accesses to this data just pull it from the cache, but my point was that I still wanted to fall back rather than fail if these properties didn't exist for whatever reason.  Given that I need to re-think my approach to pre-loading this cache, it might change the patterns a bit, but I still feel like `get_person_properties` needs to be able to call the DB in a worst case scenario so that we don't fail just because the data isn't cached.

Reply:
right makes sense 👍